internal/virter/virter.go

```diff
@@ -77,6 +80,11 @@ type PortWaiter interface {
 	WaitPort(ip net.IP, port string) error
 }
 
+// AfterNotifier wait for a duration to elapse
+type AfterNotifier interface {
```

Comment:
Another single method interface with a name that doesn't 100% fit with the go conventions. I'm open to other ideas about how to structure this.

Reply:
I find the "actualtime" package a bit weird. I mean I don't know where it is going, but `After` is just a wrapper (why not just use `time.After` directly? Everybody would know what `time.After` is, but everybody has to look up `actualtime.After`). Then it is only `Ping` which does not use/modify a common variable. Which for me, would make it a public function, not a method. But I guess we just come from different backgrounds, for me not everything has to be an object in the first place.

For the interface+mocking+testing: I'm not sure if it would not be easier to factor out the meat into smaller pure functions that get input and produce output. These would be easier to test than mocking and testing on a high(er) level. But that is then probably also a different approach to testing. And yes, this surprisingly contains many components that are nasty to test :)

Given all that, the code looks fine.

Reply:
> why not just use `time.After` directly?

Testing. I would rather use simple functions than objects, but making the tests pure of real world interactions is more important to me, and then you're stuck with interfaces to abstract away the world. Waiting for a period of time is a real world interaction.

> For the interface+mocking+testing: I'm not sure if it would not be easier to factor out the meat into smaller pure functions that get input and produce output.

It might be possible to solve it like this and it would certainly be more elegant if it worked. At the moment I suspect that there would be so little "meat" left that it would be hardly worth testing it.